### PR TITLE
[go_router] Fix parentNavigatorKey not working with multiple nested Routes in ShellRoute

### LINKED ISSUE #111961: [go_router] parentNavigatorKey not working with multiple nested Routes in ShellRoute
When a GoRoute has _rootNavigatorKey set as parentNavigatorKey, its nested routes can't be opened with "go" anymore.

## Steps to Reproduce

1. Call `GoRouter.of(context).go('/b/details/subpage');` with the routes defined as in the code below.
2. Press back
3. Press back

**Expected results:** 
1. Details-Subpage of B is shown.
2. DetailsScreen of B is shown.
3. Screen B is shown.

**Actual results:**
1. DetailsScreen for B is shown. [BAD]
2. DetailsScreen for B is shown. [Good]
3. Screen B is shown. [Good]

**Additional Information:**
If DetailsScreen has more nested routes, then for each nested screen DetailsScreen is opened once.

```
final GlobalKey<NavigatorState> _rootNavigatorKey =
    GlobalKey<NavigatorState>(debugLabel: 'root');
final GlobalKey<NavigatorState> _shellNavigatorKey =
    GlobalKey<NavigatorState>(debugLabel: 'shell');

final GoRouter _router = GoRouter(
    navigatorKey: _rootNavigatorKey,
    initialLocation: '/b',
    routes: <RouteBase>[
      /// Application shell
      ShellRoute(
        navigatorKey: _shellNavigatorKey,
        builder: (BuildContext context, GoRouterState state, Widget child) {
          return ScaffoldWithNavBar(child: child);
        },
        routes: <RouteBase>[
          GoRoute(
            path: '/b',
            builder: (BuildContext context, GoRouterState state) {
              return const ScreenB();
            },
            routes: <RouteBase>[
              GoRoute(
                path: 'details',
                parentNavigatorKey: _rootNavigatorKey,
                builder: (BuildContext context, GoRouterState state) {
                  return const DetailsScreen(label: 'B');
                },
                routes: <RouteBase>[
                  GoRoute(
                    path: 'subpage',
                    parentNavigatorKey: _rootNavigatorKey,
                    builder: (BuildContext context, GoRouterState state) {
                      return const DetailsScreen(label: 'B - subpage');
                    },
                  ),
                ],
              ),
            ],
          ),
        ],
      ),
    ],
  );
```

### PR DESCRIPTION
**Pull request by davidmigloz**
Fixes https://github.com/flutter/flutter/issues/111961

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/go_router/test/configuration_test.dart [hunk 1]
```diff
@@ -152,7 +152,7 @@ void main() {
                               builder: _mockScreenBuilder,
                               routes: <RouteBase>[
                                 GoRoute(
-                                  path: 'b',
+                                  path: 'c',
```
Comment (johnpryan) [inline packages/go_router/test/configuration_test.dart:155]: Could you add separate test for this scenario? It helps capture the API intent and helps prevent a regression from happening.

Comment (davidmigloz) [reply packages/go_router/test/configuration_test.dart]: Sounds like a better idea. I've moved it to a separate test ✅
(And move the two last tests to be inside `group('RouteConfiguration')`).

File: packages/go_router/lib/src/configuration.dart [hunk 2]
```diff
@@ -60,9 +60,9 @@ class RouteConfiguration {
               checkParentNavigatorKeys(
                 route.routes,
                 <GlobalKey<NavigatorState>>[
-                  // Once a parentNavigatorKey is used, only the navigator keys
-                  // above it can be used.
-                  ...allowedKeys.sublist(0, allowedKeys.indexOf(parentKey)),
+                  // Once a parentNavigatorKey is used, only that navigator key
+                  // or keys above it can be used.
+                  ...allowedKeys.sublist(0, allowedKeys.indexOf(parentKey) + 1),
```
Comment (chunhtai) [inline packages/go_router/lib/src/configuration.dart:65]: ah got to remember the right bound is exclusive

File: packages/go_router/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,7 @@
+## 5.1.6
+
+- Fixes `parentNavigatorKey` not working with multiple nested `GoRoute` in `ShellRoute`.
```
Comment (chunhtai) [inline packages/go_router/CHANGELOG.md:3]:
```suggestion
- Fixes crashes when multiple `GoRoute`s use the same `parentNavigatorKey` in a route subtree.
```

**Review by chunhtai — APPROVED** [review]
nice work! LGTM, %nits

**Review by johnpryan — APPROVED** [review]

**Pull request merged**